warehouse/query-core/src/main/java/datawave/query/util/sortedset/BufferedFileBackedSortedSet.java

```diff
@@ -156,56 +158,68 @@ public boolean containsAll(Collection<?> c) {
-                    if (setToCompact.getSets().size() == (setsPerCompaction - subtraction)) {
-                        newSet.addSet(compact(setToCompact));
-                        setToCompact.clear();
-                        iteration++;
-                        subtraction = (byte) (iteration < iterationsWithExtra ? 0 : 1);
-                    }
-                }
-                if (!setToCompact.isEmpty()) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Starting compaction for " + setToCompact);
-                    }
-                    long start = System.currentTimeMillis();
-                    FileSortedSet<E> compaction = compact(setToCompact);
-                    if (log.isDebugEnabled()) {
-                        long delta = System.currentTimeMillis() - start;
-                        log.debug("Compacted " + setToCompact + " -> " + compaction + " in " + delta + "ms");
-                    }
-                    newSet.addSet(compaction);
-                    setToCompact.clear();
-                }
-                this.set = newSet;
+            // create a copy of the set list (sorting below)
+            List<SortedSet<E>> sets = new ArrayList<>(set.getSets());
+            
+            // calculate the number of sets to compact
+            int numSets = sets.size();
+            int excessSets = numSets - (maxFiles / 2); // those over 50% of maxFiles
+            int setsPerCompaction = Math.min(excessSets + 1, numSets); // Add in 1 to account for the compacted set being added back in
+            
+            // sort the sets by size (compact up smaller sets first)
+            sets.sort(Comparator.comparing(SortedSet<E>::size).reversed());
+            
+            // newSet will be the final multiset
+            MultiSetBackedSortedSet<E> newSet = new MultiSetBackedSortedSet<>();
+            
+            // create a set for those sets to be compacted into one file
+            MultiSetBackedSortedSet<E> setToCompact = new MultiSetBackedSortedSet<>();
+            for (int i = 0; i < setsPerCompaction; i++) {
+                setToCompact.addSet(sets.remove(sets.size() - 1));
+            }
+            
+            // compact it
+            if (log.isDebugEnabled()) {
+                log.debug("Starting compaction for " + setToCompact);
+            }
+            long start = System.currentTimeMillis();
+            FileSortedSet<E> compaction = compact(setToCompact);
+            if (log.isDebugEnabled()) {
+                long delta = System.currentTimeMillis() - start;
+                log.debug("Compacted " + setToCompact + " -> " + compaction + " in " + delta + "ms");
+            }
+            
+            // add the compacted set to our final multiset
+            newSet.addSet(compaction);
+            
+            // clear the compactions set to remove the files that were compacted
+            setToCompact.clear();
```

Comment:
Is there some side effect to this operation? On a regular collection, I don't think this would be necessary since it's a local stack variable and all references only through it become available for GC once this method call is complete.

Reply:
The side effect is in the comment preceding: deleting the physical files for the old sets.